Fix unit_testing/test_labels.py::test_remove_missing_labels

### Description

There is an unmaintained test in test_labels.py:

https://github.com/neuropoly/spinalcordtoolbox/blob/3dfd4f81f9e4364dab4d94894ea687a625dbdea5/unit_testing/test_labels.py#L168-L170

### Steps to Reproduce

0. Check out SCT from git; and install it

1. `source python/bin/activate; conda activate venv_sct`
1. Comment out this line:
https://github.com/neuropoly/spinalcordtoolbox/blob/3dfd4f81f9e4364dab4d94894ea687a625dbdea5/unit_testing/test_labels.py#L168

2. `pytest -s -v unit_testing/test_labels.py::test_remove_missing_labels`

**Expected behavior:** <What you expect to happen>

The tests should pass.

**Actual behavior:**

The third, and longest, test fails.

```
(venv_sct) $ time pytest -s -v unit_testing/test_labels.py::test_remove_missing_labels
Status: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 7.17M/7.17M [00:01<00:00, 6.47MB/s]
Done!

================================================================ test session starts =================================================================
platform linux -- Python 3.6.12, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python
cachedir: .pytest_cache
rootdir: /home/kousu/src/neuropoly/spinalcordtoolbox, configfile: setup.cfg
plugins: forked-1.3.0, cov-2.10.1, console-scripts-1.1.0, xdist-2.2.0, cases-3.1.1
[gw0] linux Python 3.6.12 cwd: /home/kousu/src/neuropoly/spinalcordtoolbox
[gw1] linux Python 3.6.12 cwd: /home/kousu/src/neuropoly/spinalcordtoolbox
[gw0] Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56)  -- [GCC 7.3.0]
[gw1] Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56)  -- [GCC 7.3.0]
gw0 [3] / gw1 [3]
scheduling tests via LoadScopeScheduling

unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct] 
[gw0] PASSED unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct] 
unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct2] 
[gw0] PASSED unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct2] 
unit_testing/test_labels.py::test_remove_missing_labels[t2_img] 
[gw0] FAILED unit_testing/test_labels.py::test_remove_missing_labels[t2_img] 

====================================================================== FAILURES ======================================================================
_________________________________________________________ test_remove_missing_labels[t2_img] _________________________________________________________
[gw0] linux -- Python 3.6.12 /home/kousu/src/neuropoly/spinalcordtoolbox/python/envs/venv_sct/bin/python

test_remove_missing_labels_test_image = <spinalcordtoolbox.image.Image object at 0x7f4584c686d8>

>   ???

<makefun-gen-52>:2: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/envs/venv_sct/lib/python3.6/site-packages/pytest_cases/fixture_parametrize_plus.py:1023: in wrapped_test_func
    return test_func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

test_image = <spinalcordtoolbox.image.Image object at 0x7f4584c686d8>

    @parametrize("test_image", [fixture_ref(fake_3dimage_sct), fixture_ref(fake_3dimage_sct2), fixture_ref(t2_img)])
    def test_remove_missing_labels(test_image):
        src = test_image.copy()
        ref = test_image.copy()
    
        expected = test_image.copy()
    
        # introduce 2 discrepancies
        src.data[0, 0, 0] = 99
        src.data[0, 1, 2] = 99
    
        ref.data[0, 0, 0] = 1
        ref.data[0, 1, 2] = 1
    
        # manually set expected
        expected.data[0, 0, 0] = 0
        expected.data[0, 1, 2] = 0
    
        res = sct_labels.remove_missing_labels(src, ref)
        diff = res.data == expected.data
    
>       assert diff.all()
E       assert False
E        +  where False = <built-in method all of numpy.ndarray object at 0x7f45a5699b70>()
E        +    where <built-in method all of numpy.ndarray object at 0x7f45a5699b70> = array([[[False,  True,  True, ...,  True,  True,  True],\n        [ True,  True, False, ...,  True,  True,  True],\n    ...e],\n        [ True,  True,  True, ...,  True,  True,  True],\n        [ True,  True,  True, ...,  True,  True,  True]]]).all

unit_testing/test_labels.py:204: AssertionError
================================================================== warnings summary ==================================================================
unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct]
  /home/kousu/src/neuropoly/spinalcordtoolbox/unit_testing/test_labels.py:58: DeprecationWarning: get_data() is deprecated in favor of get_fdata(), which has a more predictable return type. To obtain get_data() behavior going forward, use numpy.asanyarray(img.dataobj).
  
  * deprecated from version: 3.0
  * Will raise <class 'nibabel.deprecator.ExpiredDeprecationError'> as of version: 5.0
    img = Image(i.get_data(), hdr=i.header,

unit_testing/test_labels.py::test_remove_missing_labels[fake_3dimage_sct2]
  /home/kousu/src/neuropoly/spinalcordtoolbox/unit_testing/test_labels.py:44: DeprecationWarning: get_data() is deprecated in favor of get_fdata(), which has a more predictable return type. To obtain get_data() behavior going forward, use numpy.asanyarray(img.dataobj).
  
  * deprecated from version: 3.0
  * Will raise <class 'nibabel.deprecator.ExpiredDeprecationError'> as of version: 5.0
    img = Image(i.get_data(), hdr=i.header,

unit_testing/test_labels.py::test_remove_missing_labels[t2_img]
  /home/kousu/src/neuropoly/spinalcordtoolbox/spinalcordtoolbox/image.py:350: DeprecationWarning: get_data() is deprecated in favor of get_fdata(), which has a more predictable return type. To obtain get_data() behavior going forward, use numpy.asanyarray(img.dataobj).
  
  * deprecated from version: 3.0
  * Will raise <class 'nibabel.deprecator.ExpiredDeprecationError'> as of version: 5.0
    self.data = self.im_file.get_data()

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================== short test summary info ===============================================================
FAILED unit_testing/test_labels.py::test_remove_missing_labels[t2_img] - assert False
```

--- a/spinalcordtoolbox/labels.py
+++ b/spinalcordtoolbox/labels.py
@@ -341,15 +341,12 @@ def remove_missing_labels(img: Image, ref: Image):
     :param ref: reference image
     :returns: image with labels missing from reference removed
     """
-    out = img.copy()
+    out = zeros_like(img)
 
-    input_coords = img.getNonZeroCoordinates(coordValue=True)
-    ref_coords = ref.getNonZeroCoordinates(coordValue=True)
-
-    for c in input_coords:
-        if c not in ref_coords:
-            out.data[int(c.x), int(c.y), int(c.z)] = 0
-
+    for src_coord in img.getNonZeroCoordinates(coordValue=True):
+        for ref_coord in ref.getNonZeroCoordinates(coordValue=True):
+            if src_coord == ref_coord:
+                out.data[src_coord.x, src_coord.y, src_coord.z] = src_coord.value
     return out
 
 

--- a/unit_testing/test_labels.py
+++ b/unit_testing/test_labels.py
@@ -50,6 +50,7 @@ def fake_3dimage_sct():
 
 
 test_images = [fake_3dimage_sct(), fake_3dimage_sct2(), t2_img]
+test_labels = [fake_3dimage_sct(), fake_3dimage_sct2(), labels_img]
 
 
 @pytest.mark.parametrize("test_image", test_images)
@@ -165,13 +166,10 @@ def test_compute_mse_no_label_warning(caplog):
     assert 'Label mismatch' not in caplog.text
 
 
-@pytest.mark.skip(reason="Too long to run on large image!")
-@pytest.mark.parametrize("test_image", test_images)
+@pytest.mark.parametrize("test_image", test_labels)
 def test_remove_missing_labels(test_image):
     src = test_image.copy()
     ref = test_image.copy()
-
-    expected = test_image.copy()
 
     # introduce 2 discrepancies
     src.data[0, 0, 0] = 99
@@ -180,14 +178,16 @@ def test_remove_missing_labels(test_image):
     ref.data[0, 0, 0] = 1
     ref.data[0, 1, 2] = 1
 
-    # manually set expected
-    expected.data[0, 0, 0] = 0
-    expected.data[0, 1, 2] = 0
-
     res = sct_labels.remove_missing_labels(src, ref)
-    diff = res.data == expected.data
 
-    assert diff.all()
+    res_labels = res.getNonZeroCoordinates(coordValue=True)
+    ref_labels = ref.getNonZeroCoordinates(coordValue=True)
+
+    for c in res_labels:
+        assert c in ref_labels
+
+    assert res.data[0, 0, 0] == 0
+    assert res.data[0, 1, 2] == 0
 
 
 @pytest.mark.parametrize("test_image", test_images)

